oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java

```diff
@@ -70,58 +70,93 @@ public String toString() {
-            Once we found active index, we mark all its custom version with DELETE
-            and OOTB same product version index with DELETE_HIDDEN_AND_DISABLE
-             */
-                if (indexNameObject.getProductVersion() == activeProductVersion
-                        && indexNameObject.getCustomerVersion() == 0) {
-                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObject);
-                    indexVersionOperation.setOperation(Operation.DELETE_HIDDEN_AND_DISABLE);
-                    indexVersionOperationList.add(indexVersionOperation);
-                } else {
-                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObject);
-                    indexVersionOperation.setOperation(Operation.DELETE);
-                    indexVersionOperationList.add(indexVersionOperation);
-                }
+                LOG.info("The operation for index '{}' will be: '{}'", indexName, indexVersionOperation.getOperation());
+                indexVersionOperationList.add(indexVersionOperation);
             }
         }
         if (!isValidIndexVersionOperationList(indexVersionOperationList)) {
+            LOG.info("Not valid version operation list: '{}', skip all", indexNameObjectList);
             indexVersionOperationList = Collections.emptyList();
         }
         return indexVersionOperationList;
     }
 
+    // do index purge ready based on the active index's last reindexing time is longer enough, we do this for prevent rollback
+    private static boolean isIndexPurgeReady(IndexName activeIndexName, NodeState activeIndexNode, long purgeThresholdMillis) {
+        // the 1st index must be active
+        String indexName = activeIndexName.getNodeName();
+        if (activeIndexNode.hasChildNode(IndexDefinition.STATUS_NODE)) {
+            if (activeIndexNode.getChildNode(IndexDefinition.STATUS_NODE)
+                    .getProperty(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP) != null) {
+                String reindexCompletionTime = activeIndexNode.getChildNode(IndexDefinition.STATUS_NODE)
+                        .getProperty(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP)
+                        .getValue(Type.DATE);
+                long reindexCompletionTimeInMillis = PurgeOldVersionUtils.getMillisFromString(reindexCompletionTime);
+                long currentTimeInMillis = System.currentTimeMillis();
+                if (currentTimeInMillis - reindexCompletionTimeInMillis > purgeThresholdMillis) {
+                    LOG.info("Found active index {} is longer enough", indexName);
+                    return true;
+                } else {
+                    LOG.info("The last index time '{}' isn't longer enough for: {}", reindexCompletionTime, indexName);
+                }
+            } else {
+                LOG.warn("{} property is not set for index {}", IndexDefinition.REINDEX_COMPLETION_TIMESTAMP, indexName);
```

Comment:
@thomasmueller actually not, since the job will run twice, first run it in readonly mode for getting purging list. which is especially prevent do heavy segment clone and uploading for golden publish from beginning. hence this isn't a good place to do writing. seems still be better do this during AEM init.